synchronize ref definitions

Wraith2 opened the pull request:
This PR syncs the definitions of netcore netstandard and netfx ref projects where possible. This is mainly by reordering the types into but some attributes have been added to netcore where they were present in netfx since this isn't an interface change and enhances mainly designer experiences. It's now possible to diff the ref definitions and get a pretty clear picture what the differences in public surface are between the different versions.

David-Engel commented:
I started looking at the changes and there are quite a few additions to the netcore ref file beyond the designer attributes.
- SqlCommand.NotificationAutoEnlist
- SqlConnectionStringBuilder.AsynchronousProcessing
  - (Quite a few under here actually, that have not been brought to netcore.)
I've run out of time to go through the rest. It's hard to tell what's been added through all the re-ordering. I'll try to continue on Monday.

Wraith2 commented:
removed from SqlConnectionStringBuilder.ConnectionReset, ContextConnection, NetworkLibrary, TransparentNetworkIPResolution. removed SqlCommand.NotificationAutoEnlist.
I might see if I can do something with reflection if I can find the non-ref binaries.

Wraith2 commented:
I write a program which dumps the definitions from a binary and then did some manual comparisons. interesting results overall.

[net462-ref.txt](https://github.com/dotnet/SqlClient/files/3586965/net462-ref.txt)
[netcore-impl.txt](https://github.com/dotnet/SqlClient/files/3586966/netcore-impl.txt)
[netcore-ref.txt](https://github.com/dotnet/SqlClient/files/3586967/netcore-ref.txt)
[net462-impl.txt](https://github.com/dotnet/SqlClient/files/3586968/net462-impl.txt)

David-Engel commented:
The comparisons are definitely interesting. I'm having to do a lot of research for this PR. Learning new things is not bad. Some questions in my mind are:
1. When would we want the ref definition to **not** match the implementation and why?
2. What exactly do the various designer attributes do? (I'm new to these so reading a lot but the MS docs tend to be sparse.) Is it safe to add them or is there anything that needs to be verified about methods for which we are adding those attributes?

We don't have any tests around the ref definition, so verifying this PR is all manual. It might be easier if this were split into different PRs where we can easily accept a round of reformatting/white space changes, a round of only sorting changes (no lines added or changed), and then a round of actual changes where we can ask questions about specific changes. Like why has `Microsoft.Data.SqlClient.` been prepended to a reference? Or why has `throw null;` been added or removed? Or how is it that SqlDataReader.Close() did not already exist in ref and no one noticed until now? (I guess everyone is just wrapping it in a using or calling `Dispose()`.)

Wraith2 commented:
> When would we want the ref definition to not match the implementation and why?

From my understanding they should never. They can differ with varying consequences, least damaging is the metadata. The only purpose of the ref assemblies is to be bound against by compilers when the exact TFM version has not or can not be determined. If something exists in ref but not impl then an attempt to use it will cause a runtime failure. If something exists in impl but not ref then it may or may not be visible to the compiler which could cause editor/intellisense visibility strangeness or compiler errors complaining about missing types.

The attributes I took over are all the ones that didn't need any additional references. The Security attributes are irrelevant on netcore, security is stubbed. The rest were componentmodel designer metadata which influences the propertygrid and other visual studio designer experiences, so not needed but having them present will make things feel more natural to people who use visual designers. HostProtectionAttribute was the only one that I didn't attempt to sync, that doesn't exist and won't on netcore because the reliability infrastructure it's part of is not in that runtime.

>  Like why has Microsoft.Data.SqlClient. been prepended to a reference?

Any type which has a language keyword can use that language keyword. Anything which does not should use a fully qualified name. My assumption for this is that it prevents there being any possibly namespace mistakes by virtue of there being no using directives at all, everything must be explicit about the exact type. The ref assemblies can be tool generated in corefx which is why everything is strictly in alphanumeric order.

> Or why has throw null; been added or removed?

if an attempt is made to use the ref assembly as an impl is made it should fail as soon as possible. throwing null is the pattern used throughout the other ref assemblies for this so if it was missing I added it.

>Or how is it that SqlDataReader.Close() did not already exist in ref and no one noticed until now? (I guess everyone is just wrapping it in a using or calling Dispose().)

That's what I assume. Since Close is inherited from DbDataReader it will have been present in the intellisense from the base type so even if you looked for it you'd find it and the callvirt would mean you called the right one. Not a damaging oversight. It does make it clear that the existing ref in corefx wasn't tool generated.

I can rework from the beginning but I'm not sure it'll be less confusing to do so. The process logically involves diffing the files and making repeated rounds of modifications to limit the diff output. interim points in the process have no relative value,

David-Engel commented on src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.cs:
```diff
@@ -437,70 +240,105 @@ public sealed partial class SqlCommand : System.Data.Common.DbCommand, System.IC
         public Microsoft.Data.SqlClient.SqlTransaction BeginTransaction(System.Data.IsolationLevel iso, string transactionName) { throw null; }
         public Microsoft.Data.SqlClient.SqlTransaction BeginTransaction(string transactionName) { throw null; }
         public override void ChangeDatabase(string database) { }
+        public static void ChangePassword(string connectionString, Microsoft.Data.SqlClient.SqlCredential credential, System.Security.SecureString newSecurePassword) { }
```
How come `throw null;` was removed?

Wraith2 replied:
Not sure. There were a bunch of other methods near it that didn't throw either so I've added it to a few. I'm going to see if can track down the exact rules for this since my knowledge is based on seeing ref edits and review feedback on my own.

David-Engel replied:
I am a bit confused, too, why some include it and some don't. So I'm curious what you find out. I don't feel like it's a blocker, though, given my understand of ref.

David-Engel commented on src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.cs:
```diff
@@ -513,156 +351,156 @@ public sealed partial class SqlConnection : System.Data.Common.DbConnection, Sys
         public string ApplicationName { get { throw null; } set { } }
+        [System.ComponentModel.DisplayNameAttribute("Asynchronous Processing")]
+        [System.ComponentModel.RefreshPropertiesAttribute(System.ComponentModel.RefreshProperties.All)]
+        public bool AsynchronousProcessing { get { throw null; } set { } }
```
AsynchronousProcessing does not exist in the implementation and should be removed here.

Wraith2 replied:
Removed.

David-Engel commented on src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.cs:
```diff
@@ -725,14 +565,31 @@ public partial class SqlDataReader : System.Data.Common.DbDataReader, System.IDi
         public override bool Read() { throw null; }
         public override System.Threading.Tasks.Task<bool> ReadAsync(System.Threading.CancellationToken cancellationToken) { throw null; }
+        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        System.Data.IDataReader System.Data.IDataRecord.GetData(int i) { throw null; }
```
Why would we add this here? Isn't it redundant since the class implements IDataReader?

Wraith2 replied:
The method is needed if you want to hang an attribute on it and since it's one that directly affects the visibility of this method to intellisense it seemed like a good idea.

David-Engel commented on src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.cs:
```diff
@@ -513,156 +351,156 @@ public sealed partial class SqlConnection : System.Data.Common.DbConnection, Sys
-    }
-    public delegate void SqlRowUpdatingEventHandler(object sender, SqlRowUpdatingEventArgs e);
-    public partial class SqlDataReader : System.Data.Common.DbDataReader, System.IDisposable
+    public partial class SqlDataReader : System.Data.Common.DbDataReader, System.Data.IDataReader, System.Data.IDataRecord, System.IDisposable
```
Why add `System.Data.IDataRecord`? Isn't it redundant since IDataReader implements it?

Wraith2 replied:
Yes, it's not harmful to have it but it can be removed. Done.

David-Engel commented on src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.cs:
```diff
@@ -816,22 +744,33 @@ public sealed partial class SqlParameter : System.Data.Common.DbParameter, Syste
         public override int Count { get { throw null; } }
         public override bool IsFixedSize { get { throw null; } }
         public override bool IsReadOnly { get { throw null; } }
+        public override bool IsSynchronized { get { throw null; } }
```
This doesn't appear to be in the netcore implementation.

Wraith2 replied:
Removed.

David-Engel commented on src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.cs:
```diff
@@ -816,22 +744,33 @@ public sealed partial class SqlParameter : System.Data.Common.DbParameter, Syste
-        public override void AddRange(System.Array values) { }
+        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        [System.ObsoleteAttribute("Add(String parameterName, Object value) has been deprecated.  Use AddWithValue(String parameterName, Object value).  http://go.microsoft.com/fwlink/?linkid=14202", false)]
+        public Microsoft.Data.SqlClient.SqlParameter Add(string parameterName, object value) { throw null; }
```
This doesn't appear to be in the netcore implementation.

Wraith2 replied:
Removed.

Wraith2 commented:
`SqlFacetAttribute` is in the ref but not impl. Do you want me to copy the definition from netfx into netcore? It's a surface area compatibility class since it can't possibly be useful in netcore.

David-Engel commented:
> `SqlFacetAttribute` is in the ref but not impl. Do you want me to copy the definition from netfx into netcore? It's a surface area compatibility class since it can't possibly be useful in netcore.

I don't see an impl in netfx. I think `SqlFacetAttribute` probably was meant to be removed from ref entirely when we resolved the Microsoft.SqlServer.Server namespace issue. We removed that namespace entirely and moved SqlDataRecord + dependent classes into Microsoft.Data.SqlClient.Server since SqlDataRecord seemed to be the main class used by users (we can easily bring others in if we are wrong about usage but it's hard to remove stuff).

David-Engel commented:
@Wraith2 I think this is good to merge as-is. If you want to add the `SqlFacetAttribute` change to this PR, let me know. Otherwise, I can approve it.

Wraith2 commented:
If you're happy with the current state I'd say merge it. I can come back to null throwing and the SqlFacetAttribute later in a dedicated PR if they need changes.
